### PR TITLE
A script to collect fybrik information using kubectl

### LINKED ISSUE #568: A method to collect debugging information from a running system
A method to collect debugging information from a running system

reported by:  @ronenkat 

This can be implemented as a script which will dump all relevant data to help collect information on the system.

The relevant data could contain:
- last x lines from logs of the manager and connectors,
- status of m4dapplications that are running
- dump of the relevant config maps in m4d-system

### PR DESCRIPTION
**Pull request by Vicky-Goldin**
For the record: started to work on the bash script on 28/2.

Wrote a `bash` script, which saves in a local directory the logs, output of `get pods`, modules yamls, certain configmaps, and fybrikapplication yaml if a fybrikapplication uuid (or namespace and name) were specified.
The behavior is configured by flags, which are explained in `./hack/save-logs-files.sh -h`

An example for the output of running the script from root of `fybrik`:
`./hack/save-logs-files.sh -u 1d8c95ff-0eda-45b9-a38d-291b68170ad7 -d 10m`
[fybrik_logs_2023-03-16_15-25_last_10m.zip](https://github.com/fybrik/fybrik/files/10991286/fybrik_logs_2023-03-16_15-25_last_10m.zip)

Closes #568

### CODE REVIEW

File: hack/save-logs-files.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+# Copyright 2023 IBM Corp.
+# SPDX-License-Identifier: Apache-2.0
+
+# if a uid is passed, another file will be created, containing only logs with the uid
+fybrik_application_uid=$1
```
Comment (roytman) [inline hack/save-logs-files.sh:7]: please add the command line parsing, such has https://www.baeldung.com/linux/bash-parse-command-line-arguments, and add the "help" (Usage) option

File: hack/save-logs-files.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+# variable for limiting to recent logs, e.g. 50s, 20m, 3h
+duration=1h
```
Comment (roytman) [inline hack/save-logs-files.sh:10]: another command line option

File: hack/save-logs-files.sh [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    done < $f
+done
+sort -k 3 ./$dirname_logs/combined_logs.txt > ./$dirname_logs/combined_logs_sorted.txt
+rm ./$dirname_logs/combined_logs.txt
+
+# if given fybrik application uid, create another file which only contains logs with the uid
+if [[ $fybrik_application_uid ]]; then
+    while read line; do
+        if [[ $line == *$fybrik_application_uid* ]]; then
+            echo "$line" >> ./$dirname_logs/combined_logs_sorted_onlyuid.txt
```
Comment (shlomitk1) [inline hack/save-logs-files.sh:61]: define a constant with the file name

File: hack/save-logs-files.sh [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+# variable for limiting to recent logs, e.g. 50s, 20m, 3h
+duration=1h
+
+# create a directory for saving recent logs
+dirname_logs=fybrik_logs_$(date +%Y-%m-%d_%H-%M)_last_$duration
+mkdir $dirname_logs
+mkdir $dirname_logs/by_pod
+
+# iterate over the pods in the fybrik namespaces
+for ns in $(kubectl get ns -o=jsonpath='{.items[*].metadata.name}'); do
```
Comment (shlomitk1) [inline hack/save-logs-files.sh:18]: kubectl is usually installed in hack/tools/bin. If it is not in $PATH, the executable will not be found.

**Comment by roytman** [comment]
Hi @Vicky-Goldin if your PR is ready to review, please mark it.

File: hack/save-logs-files.sh (hunk 1 above)
Comment (Vicky-Goldin) [reply hack/save-logs-files.sh:7]: done

File: hack/save-logs-files.sh (hunk 2 above)
Comment (Vicky-Goldin) [reply hack/save-logs-files.sh:10]: done

File: hack/save-logs-files.sh [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    # check if the namespace contains the word "fybrik"
+    if [[ $ns == *"fybrik"* ]]; then
```
[the remark this answers is not in the document]
Comment (Vicky-Goldin) [reply hack/save-logs-files.sh:20]: done

File: hack/save-logs-files.sh (hunk 3 above)
Comment (Vicky-Goldin) [reply hack/save-logs-files.sh:61]: done

File: hack/save-logs-files.sh [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+            fybrik_application_uuid=$2
+            shift 2
+            ;;
+        -h | --help)
+            print_description
+            print_usage
+            print_flags
+            exit 0
+            ;;
+        --) shift;
```
Comment (roytman) [inline hack/save-logs-files.sh:121]: I tried to run it on my Mac, the arguments parsing did not work.
the loop always break out here, and $1 equals to `d:n:p:u:a:h`.

Comment (Vicky-Goldin) [reply hack/save-logs-files.sh]: Looks like [in mac there is a different option of `getopt`](https://stackoverflow.com/a/7948533). It was decided that I will switch to use `getopts` which is included in `bash` but has only short flags.

File: hack/save-logs-files.sh [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+fi
+
+# create a directory for saving output
+dirname_logs=$output_path"fybrik_logs_$(date +%Y-%m-%d_%H-%M)_last_$duration"
+echo "Output will be saved in $dirname_logs"
+mkdir $dirname_logs
+dirname_logs_by_container=$dirname_logs/logs_by_container
+mkdir $dirname_logs_by_container
+
+relevant_configmaps=("cluster-metadata" "fybrik-config")
```
Comment (revit13) [inline hack/save-logs-files.sh:133]: Is it possible to print config-maps of the deployed modules in fybrik-blueprints namespace? and also the logs of the deployed modules in that namespace.

Comment (Vicky-Goldin) [reply hack/save-logs-files.sh:133]: Regarding adding configmaps , I will add to the script all the non-certificate configmaps in the explored namespaces (including `fybrik-blueprints` namespace as it is explored by default).
The logs of the pods in the `fybrik-blueprints` namespace are already extracted, including from the pods that are connected to the modules.

**Comment by Vicky-Goldin** [comment]
Closing this PR due to misuse of git commands, continuing on a new branch with a new PR: #2039

**Pull request closed without merging**